File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
import sys
fi = sys.stdin.readline
sys.setrecursionlimit(10**6)

def solution():
    global n, s, arr, cnt
    n, s = map(int, fi().split())
    arr = list(map(int, fi().split()))
    cnt = 0
    
    func(0, 0)
    if s == 0: cnt -= 1
    print(cnt)

# cur은 퀸이 있는 위치의 행의 값이다
def func(cur, tot):
    global cnt
    if cur == n:
        if tot == s: cnt += 1
        return
    
    func(cur+1, tot)
    func(cur+1, tot + arr[cur])
                
if __name__ == '__main__':
    solution()

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio cin.tie(0)->sync_with_stdio(0)
#define ll long long
#define X first
#define Y second

int n, s, cnt=0;
int arr[30];

void func(int cur, int tot) {
    if(cur == n) {
        if(tot == s) cnt++;
        return;
    }
    func(cur+1, tot); // 다음 값을 안더하고 넘어가는 경우
    func(cur+1, tot+arr[cur]); // 다음 값을 더하는 경우
}

int main(void) {
    fastio;
    // init
    cin>>n>>s;
    for(int i=0;i<n;i++) cin>>arr[i];
    
    func(0,0);
    // s==0일 때 공집합은 빼줘야 하니까 하나 빼준다
    if(s==0) cnt--;
    cout<<cnt;
    
    return 0;
}